File: leetcode/0020. Valid Parentheses/Solution.swift
// https://leetcode.com/problems/valid-parentheses/

class Solution {
    func isValid(_ s: String) -> Bool {
        let map: [Character: Character] = [")": "(", "}": "{", "]": "["]
        var stack = [Character]()
        for c in s {
            if let d = map[c] {
                let topElement = stack.isEmpty ? "#" : stack.popLast()
                if topElement != d {
                    return false
                }
            } else {
                stack.append(c)
            }
        }

        return stack.isEmpty
    }
}

let s = Solution()
assert(s.isValid("()") == true)
assert(s.isValid("()[]{}") == true)
assert(s.isValid("(]") == false)
assert(s.isValid("([)]") == false)
assert(s.isValid("{[]}") == true)

File: leetcode/0020. Valid Parentheses/Solution.py
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {")": "(", "}": "{", "]": "["}
        stack = []
        for c in s:
            if c in pairs:
                top = '?' if not stack else stack.pop()
                if top != pairs[c]:
                    return False
            else:
                stack.append(c)

        return not stack

s = Solution()
assert s.isValid("()") == True
assert s.isValid("()[]{}") == True
assert s.isValid("(]") == False
assert s.isValid("([)]") == False
assert s.isValid("{[]}") == True
